File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.py
class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        # dp = []
        # ans = 0
        # for i in range(len(nums)):
        #     row = [[0, None, None]] * len(nums)
        #     dp.append(row)
        # for i in range(len(nums)):
        #     dp[i][i] = (0, nums[i], nums[i])
        # for i in range(len(nums) - 1):
        #     for j in range(i + 1, len(nums)):
        #         currNum = nums[j]
        #         if dp[i][j-1][1] <= currNum <= dp[i][j-1][2]:
        #             dp[i][j][0] = dp[i][j-1][0]
        #             dp[i][j][1] = dp[i][j-1][1]
        #             dp[i][j][2] = dp[i][j-1][2]
        #             ans += dp[i][j][0]
        #         elif currNum > dp[i][j-1][2]:
        #             dp[i][j][0] = currNum - dp[i][j-1][1]
        #             dp[i][j][1] = dp[i][j-1][1]
        #             dp[i][j][2] = currNum
        #             ans += dp[i][j][0]
        #         else:
        #             dp[i][j][0] = dp[i][j-1][2] - currNum
        #             dp[i][j][1] = currNum
        #             dp[i][j][2] = dp[i][j-1][2]
        #             ans += dp[i][j][0]
        # return ans
        res = 0
        n = len(nums)
        for i in range(n):
            l,r = nums[i],nums[i]
            for j in range(i, n):
                l = min(l, nums[j])
                r = max(r, nums[j])
                res += r - l
        return res

File: 2104-sum-of-subarray-ranges/2104-sum-of-subarray-ranges.java
class Solution {
    public long subArrayRanges(int[] nums) {
        int n = nums.length;
        long[][][] memo = new long[n][n][2];
        for (int i = 0; i < n; i++) {
            // (MIN, MAX)
            memo[i][i][0] = nums[i];
            memo[i][i][1] = nums[i];
        }
        long ans = 0;
        for (int right = 1; right < n; right++) {
            for (int left = right - 1; left >= 0; left--) {
                long prevMin = memo[left][right - 1][0];
                long prevMax = memo[left][right - 1][1];
                long max = Math.max(nums[right], prevMax);
                long min = Math.min(nums[right], prevMin);
                ans += max - min;
                memo[left][right][0] = min;
                memo[left][right][1] = max;
            }
        }
        return ans;
    }
}